File: demo_app/app/src/main/java/com/bandyer/demo_android_sdk/notification/NotificationProxy.java
package com.bandyer.demo_android_sdk.notification;

import android.content.Context;

import com.bandyer.demo_android_sdk.utils.activities.BaseActivity;
import com.bandyer.demo_android_sdk.utils.storage.ConfigurationPrefsManager;
import com.bandyer.demo_android_sdk.utils.storage.LoginManager;

/**
 * This class is a proxy to the different notification services we want to show case
 * For the moment we implemented the FCM ( standard ) which works everywhere except for China.
 * For China we show case the Pushy library.
 */
public class NotificationProxy {

    public static final String FCM_PROVIDER = "FCM";
    public static final String PUSHY_PROVIDER = "Pushy";

    public static void registerDevice(Context context) {
        if (!LoginManager.isUserLogged(context)) return;
        String pushProvider = ConfigurationPrefsManager.getPushProvider(context);
        switch (pushProvider) {
            case FCM_PROVIDER:
                FirebaseCompat.registerDevice(context);
                break;
            case PUSHY_PROVIDER:
                PushyCompat.registerDevice(context);
                break;
        }
    }

    public static void unregisterDevice(BaseActivity context, String loggedUser) {
        String pushProvider = ConfigurationPrefsManager.getPushProvider(context);
        switch (pushProvider) {
            case FCM_PROVIDER:
                FirebaseCompat.unregisterDevice(context, loggedUser);
                break;
            case PUSHY_PROVIDER:
                PushyCompat.unregisterDevice(context, loggedUser);
                break;
        }
    }

    public static void listen(BaseActivity context) {
        String pushProvider = ConfigurationPrefsManager.getPushProvider(context);
        switch (pushProvider) {
            case PUSHY_PROVIDER:
                PushyCompat.listen(context);
                break;
        }
    }

}

File: app_utilities/src/main/java/com/bandyer/app_utilities/notification/NotificationProxy.kt
/*
 * Copyright (C) 2020 Bandyer S.r.l. All Rights Reserved.
 * See LICENSE.txt for licensing information
 */

package com.bandyer.app_utilities.notification

import android.content.Context
import com.bandyer.app_configuration.external_configuration.model.PushProvider
import com.bandyer.app_utilities.storage.ConfigurationPrefsManager
import com.bandyer.app_utilities.storage.LoginManager.isUserLogged
import com.bandyer.app_utilities.activities.BaseActivity

/**
 * This class is a proxy to the different notification services we want to show case
 * At the moment we implemented the FCM ( standard ) which works everywhere except for China.
 * For China we show case the Pushy library.
 */
object NotificationProxy {

    fun registerDevice(context: Context, loggedUser: String) {
        if (!isUserLogged(context)) return
        val configuration = ConfigurationPrefsManager.getConfiguration(context)
        when (configuration.pushProvider) {
            PushProvider.FCM -> {
                if (FirebaseCompat.isGmsAvailable(context)) FirebaseCompat.registerDevice(context, loggedUser)
                else if (HuaweiCompat.isHmsAvailable(context)) HuaweiCompat.registerDevice(context, loggedUser)
            }
            PushProvider.Pushy -> PushyCompat.registerDevice(context, loggedUser)
            PushProvider.HMS -> if (HuaweiCompat.isHmsAvailable(context)) HuaweiCompat.registerDevice(context, loggedUser)
            PushProvider.NONE -> Unit
        }
    }

    fun unregisterDevice(context: BaseActivity, loggedUser: String?) {
        val configuration = ConfigurationPrefsManager.getConfiguration(context)
        when (configuration.pushProvider) {
            PushProvider.FCM -> {
                if (FirebaseCompat.isGmsAvailable(context)) FirebaseCompat.unregisterDevice(context, loggedUser)
                else if (HuaweiCompat.isHmsAvailable(context)) HuaweiCompat.unregisterDevice(context, loggedUser)
            }
            PushProvider.Pushy -> PushyCompat.unregisterDevice(context, loggedUser)
            PushProvider.HMS -> if (HuaweiCompat.isHmsAvailable(context)) HuaweiCompat.unregisterDevice(context, loggedUser)
            PushProvider.NONE -> Unit
        }
    }

    @JvmStatic
    fun listen(context: BaseActivity) {
        val configuration = ConfigurationPrefsManager.getConfiguration(context)
        if (configuration.pushProvider != PushProvider.Pushy) return
        PushyCompat.listen(context)
    }
}